### PR TITLE
feat: add pagination base component

### PR DESCRIPTION
**Pull request by elanjm12**
Screenshot from 2022-11-08 10-19-27

### CODE REVIEW

File: src/js/base/Pagination.jsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        return pageNumbers.slice(0, 4);
+    } else if (currentPage === pageNumbers.length) {
+        return pageNumbers.slice(pageNumbers.length - 4, pageNumbers.length);
+    } else if (currentPage === pageNumbers.length - 1) {
+        return pageNumbers.slice(currentPage - 3, currentPage + 1);
+    }
+    return pageNumbers.slice(currentPage - 2, currentPage + 2);
+};
+
+export const Pagination = ({ itemsPerPage, totalItems, color }) => {
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]: Drop `color` prop. Pagination components are not going to vary in appearance over the application UI.

File: src/js/base/Pagination.jsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    if (pageNumbers.length <= 4) {
+        return pageNumbers;
+    } else if (currentPage === 1) {
+        return pageNumbers.slice(0, 4);
+    } else if (currentPage === pageNumbers.length) {
+        return pageNumbers.slice(pageNumbers.length - 4, pageNumbers.length);
+    } else if (currentPage === pageNumbers.length - 1) {
+        return pageNumbers.slice(currentPage - 3, currentPage + 1);
+    }
+    return pageNumbers.slice(currentPage - 2, currentPage + 2);
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]: Good use of `min()`, `max()`, and `range()` can greatly simplify this function.

If you are returning in `if` blocks, don't bother with `else if`. This is functionally equivalent:
```js
    if (pageNumbers.length <= 4) {
        return pageNumbers;
    } 

    if (currentPage === 1) {
        return pageNumbers.slice(0, 4);
    }

    if (currentPage === pageNumbers.length) {
        return pageNumbers.slice(pageNumbers.length - 4, pageNumbers.length);
    }

    ...        
```

File: src/js/base/Pagination.jsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        width: 100px;
+    }
+
+    button {
+        margin: 5px;
+        height: 25px;
+    }
+`;
+
+export const sliceData = (items, currentPage, itemsPerPage) => {
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]: Are you sure this is used somewhere?

File: src/js/base/Pagination.jsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return pageNumbers.slice(currentPage - 3, currentPage + 1);
+    }
+    return pageNumbers.slice(currentPage - 2, currentPage + 2);
+};
+
+export const Pagination = ({ itemsPerPage, totalItems, color }) => {
+    const [currentPage, setCurrentPage] = useState(1);
+    const paginate = number => setCurrentPage(number);
+
+    const pageNumbers = range(1, Math.ceil(totalItems / itemsPerPage + 1));
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]: Are you sure defining this array is the most efficient and clear way to implement this?

File: src/js/base/Pagination.jsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            key={pageNumber}
+            active={currentPage === pageNumber ? true : false}
+            onClick={() => paginate(pageNumber)}
+        >
+            {pageNumber}
+        </Button>
+    ));
+
+    return (
+        <StyledPaginationBox size={pageNumbers.length}>
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]: What is `size` used for?

**Review by igboyes — CHANGES_REQUESTED** [review]
For all paging endpoints we get the following:
```
{
  "found_count": 1022,
  "page": 1,
  "page_count": 41,
  "per_page": 25,
  "total_count": 1026
}
```
Keep this in mind when working on this component.

Think about where values used in your component come from. Do they belong in component state or is their authoritative source somewhere else?

File: src/js/base/Pagination.jsx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+};
+
+export const Pagination = ({ currentPage, pageCount, onPageChange }) => {
+    const paginationRange = usePagination(pageCount, currentPage);
+
+    const pageButtons = map(paginationRange, pageNumber => (
+        <Button
+            color={currentPage === pageNumber ? "blue" : ""}
+            key={pageNumber}
+            active={currentPage === pageNumber ? true : false}
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]:
```suggestion
            active={currentPage === pageNumber}
```

File: src/js/base/Pagination.jsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            {pageNumber}
+        </Button>
+    ));
+
+    return (
+        <StyledPaginationBox>
+            <Button
+                onClick={() => onPageChange(currentPage - 1)}
+                color="blue"
+                disabled={currentPage === 1 ? true : false}
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]:
```suggestion
                disabled={currentPage === 1}
```

File: src/js/base/Pagination.jsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            >
+                {"< Previous"}
+            </Button>
+            {pageButtons}
+            <Button
+                onClick={() => onPageChange(currentPage + 1)}
+                color="blue"
+                disabled={currentPage === pageCount ? true : false}
+            >
+                {"Next >"}
```
Comment (igboyes) [inline src/js/base/Pagination.jsx]:
```suggestion
                Next
```

**Review by igboyes — CHANGES_REQUESTED** [review]
Fix the bool issues and I will merge.

When we use this component, I think we are going to be forced to change it a bit. The pagination should work via URLs (eg. `?page=2`), no these page numbers and buttons will have to be links instead of having `onClick` handlers.

At that point, I'd like all of the clickable elements, except for **Next** to be styled as basic links. Next will be the most commonly clicked and should have the most visual weight.

**Comment by elanjm12** [comment]
bool issues addressed. I will address the first comment when I use this component in issue 1623.

**Review by igboyes — APPROVED** [review]

**Pull request merged**